pkg/conversation/middlewares.go

```diff
@@ -0,0 +1,211 @@
+  pass detects PII, providing both performance optimization and bug mitigation.
+*/
+
+// StreamingPIIScrubber combines window-based performance with token-aware boundaries
+// It uses a window size for performance but ensures we don't split words at the boundary
+type StreamingPIIScrubber struct {
+	buffer     []byte
+	windowSize int
+	scrubber   piiscrubber.Scrubber
+}
+
+// NewStreamingPIIScrubber creates a streaming PII scrubber with the default window size
+func NewStreamingPIIScrubber() (*StreamingPIIScrubber, error) {
+	return NewStreamingPIIScrubberWithWindowSize(PIIScrubberStreamingWindowSize)
+}
+
+// NewStreamingPIIScrubberWithWindowSize creates a streaming PII scrubber with a custom window size
+func NewStreamingPIIScrubberWithWindowSize(windowSize int) (*StreamingPIIScrubber, error) {
+	scrubber := &StreamingPIIScrubber{
+		buffer:     make([]byte, 0, windowSize*2), // 2x window size to allow for incomplete words. Reconsider using sync.Pool if this is used in a long-running/high-throughput application.
+		windowSize: windowSize,
+	}
+
+	// Initialize the default PII scrubber
+	defaultScrubber, err := piiscrubber.NewDefaultScrubber()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create default PII scrubber: %w", err)
+	}
+	scrubber.scrubber = defaultScrubber
+
+	return scrubber, nil
+}
+
+// isTokenDelimiter checks if a character is a token delimiter using Unicode categories
+func isTokenDelimiter(r rune) bool {
+	return unicode.IsSpace(r) || unicode.IsPunct(r) || unicode.IsSymbol(r)
+}
+
+// findLastCompleteWordPosition finds the position after the last complete word
+// Returns -1 if no complete word boundary found
+func findLastCompleteWordPosition(data []byte) int {
+	if len(data) == 0 {
+		return -1
+	}
+
+	// Find the last separator that creates a complete word boundary
+	for i := len(data) - 1; i >= 0; i-- {
+		if isTokenDelimiter(rune(data[i])) {
+			// Check if there's non-delimiter content after this position
+			hasContentAfter := false
+			for j := i + 1; j < len(data); j++ {
+				if !isTokenDelimiter(rune(data[j])) {
+					hasContentAfter = true
+					break
+				}
+			}
+
+			// If no content after delimiter, it's a good word boundary
+			if !hasContentAfter {
+				return i + 1
+			}
+		}
+	}
+
+	return -1
+}
+
+// ProcessChunk processes chunks using window size but respects word boundaries
+func (s *StreamingPIIScrubber) ProcessChunk(chunk []byte) ([]byte, error) {
+	if len(chunk) == 0 {
+		return chunk, nil
+	}
+
+	s.buffer = append(s.buffer, chunk...)
+
+	// Buffer until window size is reached
+	if len(s.buffer) <= s.windowSize {
+		return nil, nil
+	}
+
+	// Find the last complete word within the window
+	windowContent := s.buffer[:s.windowSize]
+	lastCompletePos := findLastCompleteWordPosition(windowContent)
+
+	if lastCompletePos == -1 {
+		// No complete words - prevent infinite buffering
+		if len(s.buffer) > s.windowSize*2 {
```

Comment:
Why are we using flushers, manual buffers etc. over actual `io` streams?

Reply:
after looking into the existing code on this, it appears that the sdk chosen does not support streaming I/O and also honestly we need to get rid of this pkg imo bc it has a bug according to the docs in this file where it fails to detect credit card numbers when IP addresses are present in the same text 💀